### PR TITLE
Add support for portable installations

### PR DESCRIPTION
**Pull request by peppy**
This requires special handling from each game which wants to use this functionality, via `GetSuitableHost`.

- Will prefer portable in all cases if a local `framework.ini` is found
- Required moving some initialisation from `GameHost`'s `ctor` to `Run()`. I think this is a good direction either way.
- [x] TODO: tests

### CODE REVIEW

File: osu.Framework/Configuration/FrameworkConfigManager.cs [hunk 1]
```diff
@@ -12,7 +12,9 @@ namespace osu.Framework.Configuration
 {
     public class FrameworkConfigManager : IniConfigManager<FrameworkSetting>
     {
-        protected override string Filename => @"framework.ini";
+        internal const string FILENAME = @"framework.ini";
+
+        protected override string Filename => FILENAME;
```
Comment (miterosan) [inline osu.Framework/Configuration/FrameworkConfigManager.cs]: What is the purpose of having the string "framework.ini" in a seperate const?

File: osu.Framework/Platform/DesktopStorage.cs [hunk 2]
```diff
@@ -54,7 +61,7 @@ public override string GetFullPath(string path, bool createIfNotExisting = false
         {
             path = path.Replace(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar);
 
-            var basePath = Path.GetFullPath(Path.Combine(BasePath, BaseName, SubDirectory));
+            var basePath = Path.GetFullPath(Path.Combine(BasePath, BaseName, SubDirectory)).TrimEnd(Path.DirectorySeparatorChar);
```
Comment (miterosan) [inline osu.Framework/Platform/DesktopStorage.cs]: Why is this needed?

File: osu.Framework/Configuration/FrameworkConfigManager.cs (hunk 1 above)
Comment (peppy) [reply osu.Framework/Configuration/FrameworkConfigManager.cs]: See the usage of it.

File: osu.Framework/Platform/DesktopStorage.cs (hunk 2 above)
Comment (peppy) [reply osu.Framework/Platform/DesktopStorage.cs]: Fails on `GetFiles(".")` with base storage folder path.

File: osu.Framework/Platform/DesktopStorage.cs [hunk 3]
```diff
@@ -13,10 +14,16 @@ public class DesktopStorage : Storage
     {
         private readonly GameHost host;
 
-        public DesktopStorage(string baseName, GameHost host)
+        public DesktopStorage(string baseName, GameHost host = null)
             : base(baseName)
         {
             this.host = host;
+
+            if ((host as DesktopGameHost)?.IsPortableInstallation == true || File.Exists(FrameworkConfigManager.FILENAME))
```
Comment (smoogipoo) [inline osu.Framework/Platform/DesktopStorage.cs]: Is there a reason why the ctor parameter isn't a `DesktopGameHost`?

File: osu.Framework/Platform/DesktopStorage.cs [hunk 4]
```diff
@@ -13,10 +14,16 @@ public class DesktopStorage : Storage
...
+            {
+                BasePath = "./";
```
Comment (smoogipoo) [inline osu.Framework/Platform/DesktopStorage.cs]: Necessary? See `LocateBasePath()`.

File: osu.Framework/Platform/GameHost.cs [hunk 5]
```diff
@@ -190,21 +188,34 @@ public double MaximumInactiveHz
...
+            });
+
+            RegisterThread(UpdateThread = new UpdateThread(UpdateFrame)
+            {
+                OnThreadStart = UpdateInitialize,
+                Monitor = { HandleGC = true },
+            });
+
+            RegisterThread(InputThread = new InputThread());
+            RegisterThread(AudioThread = new AudioThread());
```
Comment (smoogipoo) [inline osu.Framework/Platform/GameHost.cs]: What's the reason for moving this code? And more importantly what's the reason for moving it _before_ the `Logger.GameIdentifier` setter?

File: osu.Framework/Platform/DesktopGameHost.cs [hunk 6]
```diff
@@ -18,37 +18,51 @@ namespace osu.Framework.Platform
...
+                startIPC();
+
+            IsPortableInstallation = portableInstallation;
+        }
+
+        protected override void SetupForRun()
+        {
+            base.SetupForRun();
+
+            Logger.Storage = Storage.GetStorageForDirectory("logs");
```
Comment (smoogipoo) [inline osu.Framework/Platform/DesktopGameHost.cs]: This seems like it's extremely prone to failures where something in the threads would like to write to logs (stack trace collector or manual logs) because it's super late-initialised now (it seems like it would be error-prone before too, but it was in ctor so much less likely to happen).

Comment (smoogipoo) [reply osu.Framework/Platform/DesktopGameHost.cs]: On top of that, how does this work with `GameWindow`'s ctor?

File: osu.Framework/Platform/DesktopStorage.cs (hunk 3 above)
Comment (peppy) [reply osu.Framework/Platform/DesktopStorage.cs]: For iOS compatibility. Will rename this class to `NativeStorage` as discussed.

File: osu.Framework/Platform/GameHost.cs (hunk 5 above)
Comment (peppy) [reply osu.Framework/Platform/GameHost.cs]: As discussed, will rework to move all initialisation to `Run`.

File: osu.Framework/Platform/DesktopStorage.cs (hunk 4 above)
Comment (peppy) [reply osu.Framework/Platform/DesktopStorage.cs]: LocateBasePath is overridden at each implementation unfortunately.

File: osu.Framework/Platform/DesktopGameHost.cs (hunk 6 above)
Comment (peppy) [reply osu.Framework/Platform/DesktopGameHost.cs]: Will move all init to Run (didn't do this initially due to dependencies on current ordering in framework tests)
